### PR TITLE
MINOR: Fix rate metric spikes

### PR DESCRIPTION
**Pull request by emitskevich-blp**
`Rate` reports value in the form of `sumOrCount`/`monitoredWindowSize`. It has a bug in `monitoredWindowSize` calculation, which leads to spikes in result values (see real example: `io-wait-ratio` has ~100% -> ~50% spikes):
Screenshot 2024-05-07 at 10 10 33 PM

Monitored window consists of several samples (see [SampledStat](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java) for ref). Sample ends after a fixed time following the first event. New sample starts by the next event after prev sample end. There is always some gap between samples, while no event is recorded. 

Before reporting value, `Rate` purges (`sumOrCount := 0`) all samples, **started** before the monitored window (but can still overlap it). This purging is only reflected in numerator. Denominator is always the entire monitored window.

Consider the picture, corresponding to default config (`numWindows = 2`):
Screenshot 2024-05-07 at 9 48 26 PM

When measurement is requested at the end of right sample, the left sample is purged for numerator, but not for denominator. It is the reason for spikes on the chart above: when we move to such zone, numerator is reduced by 2 times, while denominator remains the same.

This PR fixes such behavior. Now samples are not purged if they overlap the monitored window, even if started before.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java [hunk 1]
```diff
@@ -64,4 +67,31 @@ public void testRateWithNoPriorAvailableSamples(int numSample, int sampleWindowS
...
+
+        // making a gap of 100 ms between windows
+        time.sleep(101);
+
+        // start the second window and record events at 0,99,199,...,999 ms
+        for (int stepMs : steps) {
+            time.sleep(stepMs);
+            rate.record(config, 1, time.milliseconds());
+            double observedRate = rate.measure(config, time.milliseconds());
+            assertTrue(observedRate >= 10 && observedRate <= 11);
```
Comment (emitskevich-blp) [inline clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java]: The last measurement here was affected by `Rate` bug and showed ~5 instead of ~10. Now it's consistent.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java [hunk 2]
```diff
@@ -106,44 +106,55 @@ public String toString() {
...
+            }
         }
+        return purgedUpToMs;
     }
 
     protected static class Sample {
         public double initialValue;
         public long eventCount;
-        public long lastWindowMs;
+        public long startTimeMs;
```
Comment (emitskevich-blp) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java:128]: It always played a role of start time. Not a functional change.

**Comment by emitskevich-blp** [comment]
@chia7712 @junrao maybe you can take a look here? Not invasive change, test failures are not related to PR changes

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java [hunk 3]
```diff
@@ -63,33 +63,38 @@ public void record(MetricConfig config, double value, long timeMs) {
...
         int minFullWindows = config.samples() - 1;
 
         // If the available windows are less than the minimum required, add the difference to the totalElapsedTime
-        if (numFullWindows < minFullWindows)
+        if (numFullWindows < minFullWindows) {
             totalElapsedTimeMs += (minFullWindows - numFullWindows) * config.timeWindowMs();
+        }
+        
+        // if some part of considered interval was just purged with its data, exclude it 
+        totalElapsedTimeMs = Math.min(totalElapsedTimeMs, now - purgedUpToMs);
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java]: This approach works for the newly added test case. However, I have two concerns.
1. It's not very repeatable. For example, if `measure()` is called a second time after a very short window, no samples will be purged and we won't be compensating for the purgedUpToMs. Then, we will fall back to use the larger window, which will result in a much lower measured value.
2. The logic is bit hard to follow now.

Here is an alternative. Note that in the new test case, the oldest window actually partially overlaps with the total sample window since the last recording falls in it. So, instead of purging it, we could keep the overlapping sample and include it in both the measure and windowSize calculation. This could make the sample window a bit larger than what's configured, but it probably provides a more accurate/consistent measurement. We need to extend `Sample` to remember the last recording time. This probably addresses both of the above concerns. What do you think?

**Review by junrao — COMMENTED** [review]
@emitskevich-blp : Thanks for identifying the issue and providing a fix. The problem is clear to me. Added a comment on the approach.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java (hunk 3 above)
Comment (emitskevich-blp) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java]: > if measure() is called a second time after a very short window, no samples will be purged and we won't be compensating for the purgedUpToMs. Then, we will fall back to use the larger window, which will result in a much lower measured value.

Thank you, I missed it

Comment (emitskevich-blp) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java]: @junrao I updated the PR according to your suggestion, works fine

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java [hunk 4]
```diff
@@ -82,7 +82,7 @@ public long windowSize(MetricConfig config, long now) {
          * but this approach does not account for sleeps. SampledStat only creates samples whenever record is called,
          * if no record is called for a period of time that time is not accounted for in windowSize and produces incorrect results.
          */
-        long totalElapsedTimeMs = now - stat.oldest(now).lastWindowMs;
+        long totalElapsedTimeMs = now - stat.oldest(now).startTimeMs;
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java:88]: Could we add some comments to note that the total elapse time could be larger than the window size since we keep the overlapping samples?

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java [hunk 5]
```diff
@@ -110,40 +111,43 @@ public String toString() {
     protected void purgeObsoleteSamples(MetricConfig config, long now) {
         long expireAge = config.samples() * config.timeWindowMs();
         for (Sample sample : samples) {
-            if (now - sample.lastWindowMs >= expireAge)
+            if (now - sample.lastEventMs >= expireAge)
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java]: Could we add a comment that we don't purge overlapping samples?

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java [hunk 6]
```diff
@@ -110,40 +111,43 @@ public String toString() {
...
+            return timeMs - startTimeMs >= config.timeWindowMs() || eventCount >= config.eventWindow();
         }
 
         @Override
         public String toString() {
             return "Sample(" +
                 "value=" + value +
                 ", eventCount=" + eventCount +
-                ", lastWindowMs=" + lastWindowMs +
+                ", startTimeMs=" + startTimeMs +
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java:156]: Should we add lastEventMs?

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java [hunk 7]
```diff
@@ -64,4 +69,30 @@ public void testRateWithNoPriorAvailableSamples(int numSample, int sampleWindowS
...
+        }
+
+        // making a gap of 100 ms between windows
+        time.sleep(101);
+
+        // start the second window and record events at 0,99,199,...,999 ms
+        for (int stepMs : steps) {
+            time.sleep(stepMs);
+            rate.record(config, 1, time.milliseconds());
+            double observedRate = rate.measure(config, time.milliseconds());
```
Comment (junrao) [inline clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java:95]: Should we do a second measurement and check that the result doesn't change?

**Review by junrao — COMMENTED** [review]
@emitskevich-blp : Thanks for the updated PR. Left a few more comments.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java [hunk 8]
```diff
@@ -72,29 +67,12 @@ public long windowSize(MetricConfig config, long now) {
...
-
-        // If the available windows are less than the minimum required, add the difference to the totalElapsedTime
-        if (numFullWindows < minFullWindows)
-            totalElapsedTimeMs += (minFullWindows - numFullWindows) * config.timeWindowMs();
-
-        // If window size is being calculated at the exact beginning of the window with no prior samples, the window size
-        // will result in a value of 0. Calculation of rate over a window is size 0 is undefined, hence, we assume the
-        // minimum window size to be at least 1ms.
-        return Math.max(totalElapsedTimeMs, 1);
+        long monitoredWindow = config.timeWindowMs() * config.samples();
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java]: Hmm, the changes the existing logic a bit. The existing logic makes sure that we include at least config.samples() - 1 full windows. The last one could be partial.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java [hunk 9]
```diff
@@ -50,10 +50,11 @@ public void record(MetricConfig config, double value, long timeMs) {
             sample = advance(config, timeMs);
         update(sample, config, value, timeMs);
         sample.eventCount += 1;
+        sample.lastEventMs = timeMs;
     }
 
     private Sample advance(MetricConfig config, long timeMs) {
-        this.current = (this.current + 1) % config.samples();
+        this.current = (this.current + 1) % (config.samples() + 1);
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java]: It would be useful to add a comment to explain why we keep an additional sample than configured.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java (hunk 4 above)
Comment (emitskevich-blp) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java:88]: Applied

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java (hunk 6 above)
Comment (emitskevich-blp) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java:156]: Applied

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java (hunk 5 above)
Comment (emitskevich-blp) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java]: Applied

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java (hunk 7 above)
Comment (emitskevich-blp) [reply clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java:95]: I’m not feeling strongly, could you please advise?

My thoughts: 
Now Rate.measure() provides repeatable results when called at/with the same timestamp. And it did so before. That non-repeatable state, previously discovered in current PR, was never committed to trunk. So it wouldn’t be clear for future readers, what’s the rationale behind this double-checking in the test.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java [hunk 10]
```diff
@@ -40,7 +40,7 @@ public abstract class SampledStat implements MeasurableStat {
 
     public SampledStat(double initialValue) {
         this.initialValue = initialValue;
-        this.samples = new ArrayList<>(2);
+        this.samples = new ArrayList<>(3);
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java]: It would be nice to add comments for this magic number :)

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java [hunk 11]
```diff
@@ -64,4 +69,31 @@ public void testRateWithNoPriorAvailableSamples(int numSample, int sampleWindowS
         double expectedRatePerSec = sampleValue / windowSize;
         assertEquals(expectedRatePerSec, observedRate, EPS);
     }
+
+    // Record an event every 100 ms on average, moving some 1 ms back or forth for fine-grained 
+    // window control. The expected rate, hence, is 10-11 events/sec depending on the moment of 
+    // measurement. Start assertions from the second window. This test is to address past issue,
+    // when measurements in the end of the sample led to value spikes.
```
Comment (junrao) [inline clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java]: How about changing "This test is to address past issue, when measurements in the end of the sample led to value spikes." to sth like "This test covers the case where a sample window partially overlaps with the monitored window." ?

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java (hunk 7 above)
Comment (junrao) [reply clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java:95]: Yes, it's probably useful to assert that taking a second measurement with no time change leads to the same value. This is more for preventing future incorrect changes and it's also low overhead.

**Review by junrao — COMMENTED** [review]
@emitskevich-blp : Thanks for the updated PR. A few more comments.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java (hunk 8 above)
Comment (emitskevich-blp) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/Rate.java]: @junrao looks like it is outdated state... I made several commits today and this line is no more in HEAD. Is there any chance you commented on intermediate state? Sorry for a little mess around commits here.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java (hunk 9 above)
Comment (emitskevich-blp) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java]: Yes, thank you, already added. Seems to be a line from intermediate state as well.

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java (hunk 11 above)
Comment (emitskevich-blp) [reply clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java]: Applied

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java (hunk 7 above)
Comment (emitskevich-blp) [reply clients/src/test/java/org/apache/kafka/common/metrics/stats/RateTest.java:95]: Applied

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java [hunk 12]
```diff
@@ -106,44 +109,51 @@ public String toString() {
...
             this.eventCount = 0;
-            this.lastWindowMs = now;
+            this.startTimeMs = now;
+            this.lastEventMs = now;
             this.value = initialValue;
         }
 
         public boolean isComplete(long timeMs, MetricConfig config) {
-            return timeMs - lastWindowMs >= config.timeWindowMs() || eventCount >= config.eventWindow();
+            return timeMs - startTimeMs >= config.timeWindowMs() || eventCount >= config.eventWindow();
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java:148]: This is unrelated to this PR, but it seems `eventCount` could be another issue if we set a non-maximum value. For example, all samples are within a single `timeWindowMs` due to a bunch of records. in this case, `totalElapsedTimeMs` will be equal to `(config.samples() - 1) * timeWindowMs`, and then we will get smaller measure due to the larger denominator

Maybe we can remove `eventWindow` as it is unused in production. This can be another PR if it is valid

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java [hunk 13]
```diff
@@ -50,10 +50,13 @@ public void record(MetricConfig config, double value, long timeMs) {
             sample = advance(config, timeMs);
         update(sample, config, value, timeMs);
         sample.eventCount += 1;
+        sample.lastEventMs = timeMs;
     }
 
     private Sample advance(MetricConfig config, long timeMs) {
-        this.current = (this.current + 1) % config.samples();
+        // need to keep one extra sample (see purgeObsoleteSamples() logic)
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java]: How about saying "we have one extra sample to remember the last recording time in order to keep the overlapping sample (see purgeObsoleteSamples() logic)"

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/SampledStatTest.java [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        stat.record(config, 1, time.milliseconds());
+        time.sleep(config.timeWindowMs() - 1);
+        stat.record(config, 1, time.milliseconds());
+        time.sleep(1);
+    }
+
+    // measure() of this impl returns the number of samples
+    static class SampleCount extends SampledStat {
+
+        SampleCount(double initialValue) {
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/metrics/stats/SampledStatTest.java]: We don't use `initialValue` in testing, so it can be simplified to 
```java
        SampleCount() {
            super(0);
        }
```

**Review by chia7712 — COMMENTED** [review]
@emitskevich-blp nice finding and fix!

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java (hunk 12 above)
Comment (junrao) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java:148]: > in this case, totalElapsedTimeMs will be equal to (config.samples() - 1) * timeWindowMs, and then we will get smaller measure due to the larger denominator

@chia7712: Could you explain your point a bit more? In the current design, we always ensure at least `(config.samples() - 1) * timeWindowMs` in `totalElapsedTimeMs`. This is to avoid the potential extreme large measured value when the accumulated sample window is small (e.g. when the metric is first recorded).

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/SampledStatTest.java (hunk 14 above)
Comment (emitskevich-blp) [reply clients/src/test/java/org/apache/kafka/common/metrics/stats/SampledStatTest.java]: Applied

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java (hunk 10 above)
Comment (emitskevich-blp) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java]: Addressed

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java (hunk 13 above)
Comment (emitskevich-blp) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java]: Addressed, but I slightly changed the suggestion, please check if it's readable. Because technically this extra space is not to remember the last recording time, it's to keep the entire extra sample with all collected data.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java (hunk 12 above)
Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java:148]: Please ignore my comments. see following test
```java
    private double test(long eventWindow) {
        Rate rate = new Rate();
        Time timer = new MockTime();
        final MetricConfig config = new MetricConfig().samples(2).timeWindow(1, SECONDS).eventWindow(eventWindow);
        // generate 9 records
        IntStream.range(0, 9).forEach(__ -> {
            rate.record(config, 1, timer.milliseconds());
            timer.sleep(1);
        });
        // almost 1 second
        timer.sleep(990);
        return rate.measure(config, timer.milliseconds());
    }
```

**before**: 3 / second
**after**: 4.5 / second

I neglect that this approach has a extra sample, and so the measure value can be larger than before.

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java (hunk 13 above)
Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java]: your comment is great! thanks!

File: clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java (hunk 12 above)
Comment (junrao) [reply clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java:148]: Yes, when the oldest window overlaps with the monitored window, the old approach ignores it and the current approach keeps it. Either one could be considered correct.

**Review by junrao — COMMENTED** [review]
@emitskevich-blp : Thanks for the updated PR. The code LGTM. There were 152 test failures. Some of them (e.g. MetricTest) seem to be related to this PR.

**Comment by chia7712** [comment]
Most failed tests can be fixed by #16044

**Review by chia7712 — APPROVED** [review]
LGTM

I can't remember the last time I seen green QA :)

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/FrequenciesTest.java [hunk 15]
```diff
@@ -89,8 +88,17 @@ public void testBooleanFrequencies() {
         }
         assertEquals(0.25, falseMetric.stat().measure(config, time.milliseconds()), DELTA);
         assertEquals(0.75, trueMetric.stat().measure(config, time.milliseconds()), DELTA);
+    }
 
-        // Record 2 more windows worth of values
+    @Test
+    public void testBooleanFrequenciesStrategy2() {
```
Comment (junrao) [inline clients/src/test/java/org/apache/kafka/common/metrics/stats/FrequenciesTest.java:94]: Does `testBooleanFrequenciesStrategy2` cover anything different from `testBooleanFrequenciesStrategy1`?

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/FrequenciesTest.java [hunk 16]
```diff
@@ -102,58 +110,69 @@ public void testBooleanFrequencies() {
...
-        assertEquals(0.00, (Double) metric3.metricValue(), DELTA);
-        assertEquals(0.50, (Double) metric4.metricValue(), DELTA);
+        assertEquals(0.25, metricValueFor("1"), DELTA);
+        assertEquals(0.25, metricValueFor("2"), DELTA);
+        assertEquals(0.00, metricValueFor("3"), DELTA);
+        assertEquals(0.50, metricValueFor("4"), DELTA);
     }
 
-    protected MetricName name(String metricName) {
+    private MetricName name(String metricName) {
```
Comment (junrao) [inline clients/src/test/java/org/apache/kafka/common/metrics/stats/FrequenciesTest.java]: name => toMetricName ?

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/FrequenciesTest.java [hunk 17]
```diff
@@ -73,14 +72,14 @@ public void testMoreFrequencyParametersThanBuckets() {
     }
 
     @Test
-    public void testBooleanFrequencies() {
+    public void testBooleanFrequenciesStrategy1() {
```
Comment (junrao) [inline clients/src/test/java/org/apache/kafka/common/metrics/stats/FrequenciesTest.java:75]: Just curious. It seems that we just break the existing tests into smaller ones. Did we change any logic for the tests to pass?

File: clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java [hunk 18]
```diff
@@ -475,28 +464,13 @@ public void testPercentiles() {
         Metric p50 = this.metrics.metrics().get(metrics.metricName("test.p50", "grp1"));
         Metric p75 = this.metrics.metrics().get(metrics.metricName("test.p75", "grp1"));
 
-        // record two windows worth of sequential values
-        for (int i = 0; i < buckets; i++)
+        // record 1-100 sequential values
+        for (int i = 0; i < buckets; i++) {
```
Comment (junrao) [inline clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java]: In some other places in this class, we don't use {} for single line statements. Should we be consistent?

File: clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java [hunk 19]
```diff
@@ -475,28 +464,13 @@ public void testPercentiles() {
...
             sensor.record(i);
-
-        assertEquals(25, (Double) p25.metricValue(), 1.0);
-        assertEquals(50, (Double) p50.metricValue(), 1.0);
-        assertEquals(75, (Double) p75.metricValue(), 1.0);
-
-        for (int i = 0; i < buckets; i++)
```
Comment (junrao) [inline clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java:486]: Why do we remove this code? Does it break the test?

**Review by junrao — COMMENTED** [review]
@emitskevich-blp : Thanks for the updated PR. A few more comments.

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/FrequenciesTest.java (hunk 17 above)
Comment (emitskevich-blp) [reply clients/src/test/java/org/apache/kafka/common/metrics/stats/FrequenciesTest.java:75]: There were 3 scenarios inside single test, which effectively should have been reset the state of metric in between. But reset was based on the knowledge of internal implementation of samples/windows, which is not a part of class API and which was slightly changed. Now it is separated to different test methods to make reset explicit.

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/FrequenciesTest.java (hunk 15 above)
Comment (emitskevich-blp) [reply clients/src/test/java/org/apache/kafka/common/metrics/stats/FrequenciesTest.java:94]: Tried just to keep all test scenarios as is to limit the scope of PR. So didn't judge it deeply. Should we?

File: clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java (hunk 19 above)
Comment (emitskevich-blp) [reply clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java:486]: Yes, it breaks the test. It relies on the knowledge of internal samples/windows implementation and effectively tests windows expiration (`SampledStat`, `SampledStatTest`) instead of testing `Percentiles`. Another option was to adapt the test for recent changes in windows expiration, but it seems better to avoid using such knowledge about internals at all. What do you think?

File: clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java (hunk 18 above)
Comment (emitskevich-blp) [reply clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java]: Applied

File: clients/src/test/java/org/apache/kafka/common/metrics/stats/FrequenciesTest.java (hunk 16 above)
Comment (emitskevich-blp) [reply clients/src/test/java/org/apache/kafka/common/metrics/stats/FrequenciesTest.java]: Tried to make now `metricName`, `freq` and `metricValue` consistent in naming, please check if it's ok

**Review by junrao — COMMENTED** [review]
@emitskevich-blp : Thanks for the updated PR. The code LGTM. Are the failed tests related to this PR, especially the following one?

`org.apache.kafka.server.ClientMetricsManagerTest.testCacheEviction()`

**Comment by emitskevich-blp** [comment]
> Are the failed tests related to this PR, especially the following one?

Not related, they are all flaky tests

**Review by junrao — COMMENTED** [review]
@emitskevich-blp : Thanks for verifying the failed tests. The PR LGTM. Thanks for your contribution!
